### PR TITLE
Add IDP_ID to the tables IDN_OAUTH2_ACCESS_TOKEN and IDN_OAUTH2_AUTHORIZATION_CODE

### PR DESCRIPTION
**Pull request by DMHP**
### Proposed changes in this pull request
Introduce IDP name to oauth tables
Add IDP_ID related changes to authz code queries and DAO
Add idpName param for createAuthenticatedUser
Modify Unit tests

### When should this PR be merged
Immediately

### Follow up actions
None

### CODE REVIEW

**Comment by janakamarasena** [comment]
This PR addresses https://github.com/wso2/product-is/issues/4879

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/OAuthUtil.java [hunk 1]
```diff
@@ -74,16 +74,43 @@ public static void clearOAuthCache(String consumerKey, User authorizedUser) {
...
         user = UserCoreUtil.addTenantDomainToEntry(user, authorizedUser.getTenantDomain());
-        clearOAuthCache(consumerKey, user, scope);
+        String tenantDomain = authorizedUser.getTenantDomain();
+        String authenticatedIDP;
+        if (authorizedUser instanceof AuthenticatedUser) {
+            authenticatedIDP = ((AuthenticatedUser) authorizedUser).getFederatedIdPName();
+        } else {
+            authenticatedIDP = null;
+            if (log.isDebugEnabled()) {
+                log.debug("User is not an instance of AuthenticatedUser therefore cannot resolve authenticatedIDP " +
```
Comment (vihanga-liyanage) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/OAuthUtil.java]: suggestion: 
```User object is not an instance of AuthenticatedUser. Therefore cannot resolve authenticatedIDP...```

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/AccessTokenDAOImpl.java [hunk 2]
```diff
@@ -480,54 +496,89 @@ private AccessTokenDO getLatestAccessTokenByState(Connection connection, String
...
+                    if (connection.getMetaData().getDriverName().contains("MySQL")
+                            || connection.getMetaData().getDriverName().contains("H2")) {
+                        sql = SQLQueries.RETRIEVE_LATEST_NON_ACTIVE_ACCESS_TOKEN_BY_CLIENT_ID_USER_SCOPE_IDP_NAME_MYSQL;
+                    } else if (connection.getMetaData().getDatabaseProductName().contains("DB2")) {
+                        sql = SQLQueries.RETRIEVE_LATEST_NON_ACTIVE_ACCESS_TOKEN_BY_CLIENT_ID_USER_SCOPE_IDP_NAME_DB2SQL;
+                    } else if (connection.getMetaData().getDriverName().contains("MS SQL")
+                            || connection.getMetaData().getDriverName().contains("Microsoft")) {
+                        sql = SQLQueries.RETRIEVE_LATEST_NON_ACTIVE_ACCESS_TOKEN_BY_CLIENT_ID_USER_SCOPE_IDP_NAME_MSSQL;
+                    } else if (connection.getMetaData().getDriverName().contains("PostgreSQL")) {
+                        sql = SQLQueries.RETRIEVE_LATEST_NON_ACTIVE_ACCESS_TOKEN_BY_CLIENT_ID_USER_SCOPE_IDP_NAME_POSTGRESQL;
```
Comment (vihanga-liyanage) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/AccessTokenDAOImpl.java]: Character count exceeds 120 limit.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java [hunk 3]
```diff
@@ -567,6 +569,7 @@ public static String buildCacheKeyStringForAuthzCode(String clientId, String aut
      * @param authorizedUser
      * @return
      */
+    @Deprecated
```
Comment (vihanga-liyanage) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java]: Please add a reason to deprecate and the new method to be used instead.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java [hunk 4]
```diff
@@ -2702,6 +2734,7 @@ public static String getFederatedIdPFromDomain(String userStoreDomain) {
      * @param tenantDomain    tenent domain
      * @return an instance of AuthenticatedUser{@link AuthenticatedUser}
      */
+    @Deprecated
```
Comment (vihanga-liyanage) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java]: Please add a new method to be used instead

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java [hunk 5]
```diff
@@ -2722,6 +2755,52 @@ public static AuthenticatedUser createAuthenticatedUser(String username, String
...
+                authenticatedUser.setFederatedIdPName(idpName);
+            } else {
+                authenticatedUser.setUserStoreDomain(userStoreDomain);
+            }
+        } else {
+            // Preserving behaviour in deprecated createAuthenticatedUser method which did not have the idpName.
+            if (StringUtils.startsWith(userStoreDomain, OAuthConstants.UserType.FEDERATED_USER_DOMAIN_PREFIX) &&
+                    !OAuthServerConfiguration.getInstance().isMapFederatedUsersToLocal()) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Federated prefix found in domain: " + userStoreDomain + " for user: " + username + " in " +
```
Comment (vihanga-liyanage) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java]: Let's keep the charactor count to 120 per line.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/OAuthUtil.java (hunk 1 above)
Comment (janakamarasena) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/OAuthUtil.java]: Fixed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1128

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/AccessTokenDAOImpl.java (hunk 2 above)
Comment (janakamarasena) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/AccessTokenDAOImpl.java]: Fixed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1128

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java (hunk 3 above)
Comment (janakamarasena) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java]: Fixed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1128

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java (hunk 4 above)
Comment (janakamarasena) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java]: Fixed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1128

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java (hunk 5 above)
Comment (janakamarasena) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java]: Fixed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1128

**Pull request merged**